base/iterators.jl

```diff
@@ -1202,7 +1203,19 @@ julia> [(x,y) for x in 0:1 for y in 'a':'c']  # collects generators involving It
 flatten(itr) = Flatten(itr)
 
 eltype(::Type{Flatten{I}}) where {I} = eltype(eltype(I))
-eltype(::Type{Flatten{I}}) where {I<:Union{Tuple,NamedTuple}} = promote_typejoin(map(eltype, fieldtypes(I))...)
+
+# For tuples, we statically know the element type of each index, so we can compute
+# this at compile time.
+function eltype(::Type{Flatten{I}}) where {I<:Union{Tuple,NamedTuple}}
+    @_foldable_meta
+    T = Union{}
+    for i in fieldtypes(I)
+        T = Base.promote_typejoin(T, eltype(i))
+        T === Any && return Any
```

Comment:
```suggestion
```

This line is unnecessary, right?

Reply:
Yes, but we might as well have Julia do less work at compile time where possible. This fast-path will probably be hit often.